### ISSUE TITLE
Question on the term "pretrained BertSum"

### ISSUE THREAD

**Issue by robinsongh381** [nlpyang/PreSumm#33]
On your paper, you mentioned "The encoder is pretrained BertSum" and im not entirely sure on the meaning of "pretrained BertSum".

For an input document, you modify BERT as follows:
(1) Add CLS tokens in the beginning of each sentence
(2) interval segment embeddings

This leads to BertSum.

On what task(s), do u perform and call the trained model "pretrained BertSum" ?

Just MLM task ? or borth MLM and NSP just like in BERT pretrain ?

Thanks in advance

**Comment by huseinzol05** [comment]
BERTSUM just a feeding technique, i am serious, I study the code to translate to Tensorflow.

![Screenshot 2019-09-23 at 5 12 52 PM](https://user-images.githubusercontent.com/19810909/65414007-5acc3780-de25-11e9-81f3-ba24457db41d.png)

**Comment by robinsongh381** [comment]
Are you sure ?

If you just change the way of constructing an input (i.e/ (1) Add CLS tokens in the beginning of each  and (2) interval segment embeddings) and use the pretrained BERT whose input involves in only one CLS token for each input, how do we get meaningful output ?

I thought that we need to "pretrain again" with the BERTSUM feeding method against MLM and NSP ?!

What do u think @huseinzol05 ?

**Comment by huseinzol05** [comment]
Nope, no need. It is literally just constructing the inputs. I reversed engineered the code to Tensorflow already.

**Comment by robinsongh381** [comment]
@huseinzol05 
I am surprised !
Is there any chance that you could possibly share the tensorflow code ?
That would be greatly helpful for my understanding

**Comment by huseinzol05** [comment]
This is preprocessing and tokenization, https://github.com/huseinzol05/NLP-Models-Tensorflow/blob/master/extractive-summarization/preprocessing-data-bert.ipynb

This is the model, https://github.com/huseinzol05/NLP-Models-Tensorflow/blob/master/extractive-summarization/4.bert-base.ipynb

I got bad result for now,

1. I disable `use_position_embeddings` inside https://github.com/huseinzol05/NLP-Models-Tensorflow/blob/master/extractive-summarization/modeling.py#L190

Original implementation is using `use_position_embeddings`. Problem if we implemented `use_position_embeddings`, max length BERT can accept is 512 length, and obviously, some texts we to summarize longer than that.

If we check how nlpyang/PreSumm tackle this issue, https://github.com/nlpyang/PreSumm/blob/master/src/models/model_builder.py#L150 , nlpyang/PreSumm repeat after 512, and that pytorch code is unable to done it tensorflow.

**Comment by robinsongh381** [comment]
Thanks for your reply.

Can we still use 'use_position_embeddings' , with the input sequence length greater than 512 ?

**Comment by huseinzol05** [comment]
`Can we still use use_position_embeddings, with the input sequence length greater than 512 ?
In my opinion, 512 is just a hyper-parameter and we can simply change it to another value.` 
We can't, `use_position_embeddings` in Tensorflow code will exception if sequence longer than 512.

`As you said, if BertSum is just a way of modifying input, rather than a new way of pre-training, I don't see that the former definition of BertSum can capture the true meaning of CLS token.`.
It will learn during transfer learning extractive / abstractive summarization. We know original BERT can capture 2 sentences by putting `CLS` token, example, text similarity transfer learning. BERTSUM just added `CLS` after N sentences, and during transfer learning, we give indices of `CLS` to let know BERTSUM when to gather output sequences to multiple output N sequences depends on `CLS` counts.

**Comment by zakhan4** [comment]
Hi,

I have a similar understanding as @robinsongh381. 
Can anyone please provide a satisfying answer? Thanks!
